### PR TITLE
User API v2 Add to Cart API endpoint instead of OrdersController#populate

### PR DESCRIPTION
**Pull request by damianlegawiec**

### CODE REVIEW

File: frontend/app/assets/javascripts/spree/frontend/product.js [hunk 1]
```diff
@@ -78,3 +80,49 @@ Spree.ready(function ($) {
 
   return Spree.addImageHandlers()
 })
+
+Spree.ready(function ($) {
```
Comment (devilcoders) [inline frontend/app/assets/javascripts/spree/frontend/product.js]: No jQuery 🎉 is used in this function, so no need for `$`
```suggestion
Spree.ready(() => {
```

File: frontend/app/assets/javascripts/spree/frontend/product.js [hunk 2]
```diff
@@ -78,3 +80,49 @@ Spree.ready(function ($) {
...
+      // another one
+      SpreeAPI.Storefront.ensureCart(
+        function () {
+          SpreeAPI.Storefront.addToCart(
+            variantId,
+            quantity,
+            {}, // options hash - you can pass additional parameters here, your backend
+                // needs to be aware of those, see API docs: https://github.com/spree/spree/blob/master/api/docs/v2/storefront/index.yaml#L42
+            function () {
+              // redirect wity `variant_id` is crucial for analytics tracking
```
Comment (devilcoders) [inline frontend/app/assets/javascripts/spree/frontend/product.js]: Typo
```suggestion
              // redirect with `variant_id` is crucial for analytics tracking
```

File: frontend/app/assets/javascripts/spree/frontend/product.js [hunk 3]
```diff
@@ -78,3 +80,49 @@ Spree.ready(function ($) {
...
+              // provided by `spree_analytics_trackers` extension
+              window.location = Spree.routes.cart + '?variant_id=' + variantId.toString()
```
Comment (devilcoders) [inline frontend/app/assets/javascripts/spree/frontend/product.js]:
```suggestion
              window.location = `${Spree.routes.cart}?variant_id=${variantId.toString()}`
```

**Review by oleksify — CHANGES_REQUESTED** [review]

File: api/app/assets/javascripts/spree/api/storefront/cart.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+SpreeAPI.Storefront.createCart = function (successCalback, failureCallback, setCookie = false) {
+  var headers = {}
+  // if signed in we need to pass the Bearer authorization token
+  // to assign this newly created Order to the currently signed in user
+  if (SpreeAPI.oauthToken) {
+    headers['Authorization'] = 'Bearer ' + SpreeAPI.oauthToken
+  }
+
+  // setting this cookie is essential for communication with Rails frontend
+  var createCartUrl = Spree.routes.api_v2_storefront_cart_create
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply api/app/assets/javascripts/spree/api/storefront/cart.js]: @devilcoders this is still built with asset pipeline so no ES6 support :(

**Review by pl-spark — CHANGES_REQUESTED** [review]

File: api/app/controllers/spree/api/v2/storefront/cart_controller.rb [hunk 5]
```diff
@@ -3,6 +3,8 @@ module Api
     module V2
       module Storefront
         class CartController < ::Spree::Api::V2::BaseController
+          include ActionController::Cookies
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply api/app/controllers/spree/api/v2/storefront/cart_controller.rb]: @piotrleniec-spark it's required for the default rails frontend , we don't read cookies anywhere in API v2, authorization is based on API tokens

[the remark this answers is not in the document]
Comment (damianlegawiec) [reply api/app/controllers/spree/api/v2/storefront/cart_controller.rb]: @piotrleniec-spark fixed via https://github.com/spree/spree/pull/9050/commits/5fbc8d48742ebd08a1792c6c14725e07511fea0a

File: frontend/app/assets/javascripts/spree/frontend/cart.js [hunk 6]
```diff
@@ -37,3 +37,23 @@ Spree.fetch_cart = function () {
...
+    successCallback()
+  } else {
+    fetch(Spree.routes.ensure_cart, {
+      method: 'POST',
+      credentials: 'same-origin'
+    }).then(function (response) {
+      switch (response.status) {
+        case 200:
+          response.json().then(function (json) {
+            SpreeAPI.orderToken = json.token
```
Comment (piotrleniec-spark) [inline frontend/app/assets/javascripts/spree/frontend/cart.js]: @damianlegawiec, maybe we could also store that token in `localStorage`?

**Review by pl-spark — APPROVED** [review]

**Review by oleksify — APPROVED** [review]

**Pull request merged**